### PR TITLE
[testing] improve broadcasts_input error message

### PR DESCRIPTION
**Pull request by kshitij12345**
Context:
The Error message when `broadcasts_input` is marked incorrectly is uninformative [See Error Currently]
https://github.com/pytorch/pytorch/pull/57941#discussion_r631749435

Error Currently
```
Traceback (most recent call last):
  File "/home/kshiteej/Pytorch/pytorch_i0_promotion/test/test_ops.py", line 326, in test_variant_consistency_eager
    _test_consistency_helper(samples, variants)
  File "/home/kshiteej/Pytorch/pytorch_i0_promotion/test/test_ops.py", line 310, in _test_consistency_helper
    variant_forward = variant(cloned,
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/unittest/case.py", line 227, in __exit__
    self._raiseFailure("{} not raised".format(exc_name))
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/unittest/case.py", line 164, in _raiseFailure
    raise self.test_case.failureException(msg)
AssertionError: RuntimeError not raised
```

Error After PR
```
Traceback (most recent call last):
  File "/home/kshiteej/Pytorch/pytorch_i0_promotion/test/test_ops.py", line 329, in test_variant_consistency_eager
    _test_consistency_helper(samples, variants)
  File "/home/kshiteej/Pytorch/pytorch_i0_promotion/test/test_ops.py", line 313, in _test_consistency_helper
    variant_forward = variant(cloned,
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/unittest/case.py", line 227, in __exit__
    self._raiseFailure("{} not raised".format(exc_name))
  File "/home/kshiteej/.conda/envs/pytorch-cuda-dev/lib/python3.8/unittest/case.py", line 164, in _raiseFailure
    raise self.test_case.failureException(msg)
AssertionError: RuntimeError not raised : inplace variant either allowed resizing or you have marked the sample SampleInput(input=Tensor, args=(tensor([[[ 2.1750, -8.5027, -3.1403, -6.9942,  3.2609],
         [-2.5057, -5.9123, -5.4633,  6.1203, -8.2124],
         [-3.5802, -8.4869, -6.0700,  2.3431, -8.1955],
         [-7.3316,  1.3248, -6.8661,  7.1483, -8.0719],
         [ 4.5977, -4.0448, -6.2044, -2.1314, -8.4956]],

        [[ 3.2769, -8.4360,  1.2826,  7.1749,  4.7653],
         [-0.2816, -2.5997, -4.7659, -3.7814,  3.9704],
         [-2.1778, -3.8117, -6.0276, -0.8423, -5.9646],
         [ 8.6544, -3.0922,  0.2558, -4.9318, -4.7596],
         [ 4.5583,  4.3830,  5.8793,  0.9713, -2.1481]],

        [[-1.0447,  0.9334,  7.6405, -4.8933, -7.4010],
         [ 7.7168, -8.4266, -5.5980, -6.9368,  7.1309],
         [-8.7720, -5.0890, -0.4975,  1.9518,  1.7074],
         [-8.5783,  8.5510, -8.5459, -3.5451,  8.4319],
         [ 8.5052, -8.9149, -6.6298, -1.2750, -5.7367]],

        [[-6.5625,  8.2795, -4.9311,  1.9501, -7.1777],
         [-8.4035,  1.1136, -7.6418, -7.0726, -2.8281],
         [ 4.2668, -0.2883, -6.2246,  2.3396,  1.2911],
         [ 4.6550, -1.9525,  4.4873, -3.8061, -0.8653],
         [-3.4256,  4.4423,  8.2937, -5.3456, -4.2624]],

        [[ 7.6128, -6.3932,  4.7131, -5.4938,  6.4792],
         [-6.5385,  2.4385,  4.5570,  3.7803, -8.3281],
         [-2.9785, -4.4745, -1.1778, -8.9324,  1.3663],
         [ 3.7437,  3.5171, -6.3135, -8.4519, -2.7033],
         [-5.0568, -8.4630, -4.2870, -3.7284, -1.5238]]], device='cuda:0',
       dtype=torch.float32, requires_grad=True),), broadcasts_input=True) incorrectly with `broadcasts_self=True
```

**NOTE**: 
Printing the sample looks very verbose and it may be hard to figure out which sample is incorrectly configured if there are multiple samples with similar input shapes.

Two Options to make this error less verbose
* Don't print the sample and just print `inplace variant either allowed resizing or you have marked one of the sample incorrectly with broadcasts_self=True`
* Have some mechanism to name samples which will be printed in the `repr` (which will need extra machinery)

### CODE REVIEW

File: test/test_ops.py [hunk 1]
```diff
@@ -306,7 +306,10 @@ def _test_consistency_helper(samples, variants):
                     cloned = clone_input_helper(sample.input) if variant in inplace_ops else sample.input
 
                     if variant in inplace_ops and sample.broadcasts_input:
-                        with self.assertRaises(RuntimeError):
+                        with self.assertRaises(RuntimeError,
+                                               msg=('inplace variant either allowed '
```
Comment (mruberry) [inline test/test_ops.py:310]: Language looks good overall, suggested tweak:

"inplace variant either incorrectly allowed..."

Just to confirm this is a bug if it occurs

**Comment by mruberry** [comment]
This is a cool idea; can we improve the printing of the SampleInputs while we're here?

https://github.com/pytorch/pytorch/blob/71f4c5c1f436258adc303b710efb3f41b2d50c4e/torch/testing/_internal/common_methods_invocations.py#L106

There are two changes I'd like to propose:

- change dunder repr to not shorthand the input but print it like args for consistency
- create a new method called something like `summary()` that:
  - prints all tensors as Tensor[shape], so a 4x3 matrix would be Tensor[4, 3]
  - prints all TensorLists as TensorList[Tensor[shape], Tensor[shape], ...], so a list of two tensors might be TensorList[Tensor[1, 3], Tensor[3, 1]], for example
  - prints everything else as usual

Dunder repr should be returning something code-like (that is, if its result is copy-pasted into Python it should probably produce a SampleInput), and it's weird that we only "summarize" or "abstract" input. This second type of printing should look much nicer in this case and help people identify the sample input more easily.

I agree with you that naming sample inputs would extend both their dunder reprs and "summaries" or "abstract representations" in a unique way. If you'd like to extend sample inputs with optional names then let's do it. It can be part of this PR if you like.

How does that sound?

**Comment by kshitij12345** [comment]
Error after the new summary update:
```
AssertionError: RuntimeError not raised : inplace variant either incorrectly allowed resizing or you have marked the sample SampleInput(input=Tensor[5, 5, 5], args=TensorList[Tensor[5, 5, 5]], kwargs={}, output_process_fn_grad=None, broadcasts_input=True, name='') incorrectly with `broadcasts_self=True
```

Tested with

<details>

```python
from torch.testing._internal.common_methods_invocations import (SampleInput, make_tensor)
import torch
from functools import partial

make_arg = partial(make_tensor, device='cpu', dtype=torch.float, requires_grad=True)

print("Scalar Tensor", SampleInput(make_arg(())).summary(), "\n")

print("2-D Tensor", SampleInput(make_arg((2, 3))).summary(), "\n")

inp = [make_arg((2, 3)), make_arg((2, 3)), make_arg((2, 3))]

print("Tensor List", SampleInput(inp).summary(), "\n")

lst = [make_arg((3, 3)), make_arg((2, 3)), make_arg((1, 3))]

print("kwargs with tensor list", SampleInput(inp, kwargs={"other": make_arg((3, 3)), "t_list": lst}).summary(), "\n")

print("Args with 2 tensors and Python int", SampleInput(inp, args=(
    make_arg((3, 4)), make_arg((4, 3)), 0), name="test_sample").summary(), "\n")


print("Args tuple with tensor list and name", SampleInput(inp, args=(lst, 0), name="test_sample").summary(), "\n")

print("Args list  with tensor list, kwargs and name", SampleInput(inp, args=[lst, 0], kwargs={"other": make_arg(
    (3, 3)), "t_list": lst}, name="test_sample").summary(), "\n")

```

</details>

Output
<details>

```
Scalar Tensor SampleInput(input=Tensor[], args=(), kwargs={}, output_process_fn_grad=None, broadcasts_input=False, name='') 

2-D Tensor SampleInput(input=Tensor[2, 3], args=(), kwargs={}, output_process_fn_grad=None, broadcasts_input=False, name='') 

Tensor List SampleInput(input=TensorList[Tensor[2, 3], Tensor[2, 3], Tensor[2, 3]], args=(), kwargs={}, output_process_fn_grad=None, broadcasts_input=False, name='') 

kwargs with tensor list SampleInput(input=TensorList[Tensor[2, 3], Tensor[2, 3], Tensor[2, 3]], args=(), kwargs={'other': 'Tensor[3, 3]', 't_list': 'TensorList[Tensor[3, 3], Tensor[2, 3], Tensor[1, 3]]'}, output_process_fn_grad=None, broadcasts_input=False, name='') 

Args with 2 tensors and Python int SampleInput(input=TensorList[Tensor[2, 3], Tensor[2, 3], Tensor[2, 3]], args=(Tensor[3, 4],Tensor[4, 3],0), kwargs={}, output_process_fn_grad=None, broadcasts_input=False, name='test_sample') 

Args tuple with tensor list and name SampleInput(input=TensorList[Tensor[2, 3], Tensor[2, 3], Tensor[2, 3]], args=(TensorList[Tensor[3, 3], Tensor[2, 3], Tensor[1, 3]],0), kwargs={}, output_process_fn_grad=None, broadcasts_input=False, name='test_sample') 

Args list  with tensor list, kwargs and name SampleInput(input=TensorList[Tensor[2, 3], Tensor[2, 3], Tensor[2, 3]], args=(TensorList[Tensor[3, 3], Tensor[2, 3], Tensor[1, 3]],0), kwargs={'other': 'Tensor[3, 3]', 't_list': 'TensorList[Tensor[3, 3], Tensor[2, 3], Tensor[1, 3]]'}, output_process_fn_grad=None, broadcasts_input=False, name='test_sample') 
```

</details>

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -103,17 +104,47 @@ def __init__(self, input, *, args=tuple(), kwargs=None, output_process_fn_grad=N
         # for such inputs (as they will error out otherwise).
         self.broadcasts_input = broadcasts_input
 
-    def __repr__(self):
+    def _repr_helper(self, formatter):
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:107]: Add comments for each of these functions explaining what they do and their intended use

**Review by mruberry — APPROVED** [review]
Cool! Nice work, @kshitij12345. Just needs a few comments to help explain what's up with the code if anyone stumbles into this

Ping me when this is ready to merge

**Pull request closed without merging**